module/icp/algs/aes/aes_impl.c

```diff
@@ -1567,52 +219,214 @@ aes_alloc_keysched(size_t *size, int kmflag)
+const aes_impl_ops_t *aes_all_impl[] = {
+	&aes_generic_impl,
+#if defined(__x86_64)
+	&aes_x86_64_impl,
+#endif
+#if defined(__x86_64) && defined(HAVE_AES)
+	&aes_aesni_impl,
+#endif
+};
+
+/* Indicate that benchmark has been completed */
+static boolean_t aes_impl_initialized = B_FALSE;
+
+/* Select aes implementation */
+#define	IMPL_FASTEST	(UINT32_MAX)
+#define	IMPL_CYCLE	(UINT32_MAX-1)
 
-#ifdef __amd64
+#define	AES_IMPL_READ(i) (*(volatile uint32_t *) &(i))
 
-#define	INTEL_AESNI_FLAG (1 << 25)
+static uint32_t icp_aes_impl = IMPL_FASTEST;
+static uint32_t user_sel_impl = IMPL_FASTEST;
+
+/* Hold all supported implementations */
+static size_t aes_supp_impl_cnt = 0;
+static aes_impl_ops_t *aes_supp_impl[ARRAY_SIZE(aes_all_impl)];
 
 /*
- * Return 1 if executing on Intel with AES-NI instructions,
- * otherwise 0 (i.e., Intel without AES-NI or AMD64).
- * Cache the result, as the CPU can't change.
+ * Selects the aes operations for encrypt/decrypt/key setup
  */
-static int
-intel_aes_instructions_present(void)
+aes_impl_ops_t *
+aes_impl_get_ops()
 {
-	static int cached_result = -1;
-	unsigned eax, ebx, ecx, edx;
-	unsigned func, subfunc;
+	aes_impl_ops_t *ops = NULL;
+	const uint32_t impl = AES_IMPL_READ(icp_aes_impl);
 
-	if (cached_result == -1) { /* first time */
-		/* check for an intel cpu */
-		func = 0;
-		subfunc = 0;
+	switch (impl) {
+	case IMPL_FASTEST:
+		ASSERT(aes_impl_initialized);
+		ops = &aes_fastest_impl;
+		break;
+	case IMPL_CYCLE:
+	{
+		ASSERT(aes_impl_initialized);
+		ASSERT3U(aes_supp_impl_cnt, >, 0);
+		/* Cycle through supported implementations */
+		static size_t cycle_impl_idx = 0;
+		size_t idx = (++cycle_impl_idx) % aes_supp_impl_cnt;
+		ops = aes_supp_impl[idx];
+	}
+	break;
+	default:
+		ASSERT3U(impl, <, aes_supp_impl_cnt);
+		ASSERT3U(aes_supp_impl_cnt, >, 0);
+		if (impl < ARRAY_SIZE(aes_all_impl))
+			ops = aes_supp_impl[impl];
+		break;
+	}
 
-		__asm__ __volatile__(
-		    "cpuid"
-		    : "=a" (eax), "=b" (ebx), "=c" (ecx), "=d" (edx)
-		    : "a"(func), "c"(subfunc));
+	ASSERT3P(ops, !=, NULL);
 
-		if (memcmp((char *)(&ebx), "Genu", 4) == 0 &&
-		    memcmp((char *)(&edx), "ineI", 4) == 0 &&
-		    memcmp((char *)(&ecx), "ntel", 4) == 0) {
-			func = 1;
-			subfunc = 0;
+	return (ops);
+}
 
-			/* check for aes-ni instruction set */
-			__asm__ __volatile__(
-			    "cpuid"
-			    : "=a" (eax), "=b" (ebx), "=c" (ecx), "=d" (edx)
-			    : "a"(func), "c"(subfunc));
+void
+aes_impl_init(void)
+{
+	aes_impl_ops_t *curr_impl;
+	int i, c;
 
-			cached_result = !!(ecx & INTEL_AESNI_FLAG);
-		} else {
-			cached_result = 0;
+	/* move supported impl into aes_supp_impls */
+	for (i = 0, c = 0; i < ARRAY_SIZE(aes_all_impl); i++) {
+		curr_impl = (aes_impl_ops_t *)aes_all_impl[i];
+
+		if (curr_impl->is_supported())
+			aes_supp_impl[c++] = (aes_impl_ops_t *)curr_impl;
+	}
+	aes_supp_impl_cnt = c;
+
+	/* set fastest implementation. assume hardware accelerated is fastest */
```

Comment:
Based on the local testing I've done with this change it seems we can't assume the hardware accelerated version will always be fastest.  Using a dual-socket "AMD EPYC 7451 24-Core Processor" I was getting the best performance with `icp_aes_impl=x86_64` followed by the `generic` and then the `aesni` version was bringing up the rear at a little better than _half_ the x86_64 rate.

@Teknoman117 can you look in to adding a micro-benchmark similar to the RAIDZ code to dynamically determine the fastest implementation during initiailization.